avr-hal-generic/src/adc.rs

```diff
@@ -0,0 +1,165 @@
+                        ClockRateDivision::Factor64 => w.adps().prescaler_64(),
+                        ClockRateDivision::Factor128 => w.adps().prescaler_128(),
+                    }});
+                self.peripheral.admux.write(|w| match settings.ref_voltage {
+                    ReferenceVoltage::Aref => w.refs().aref(),
+                    ReferenceVoltage::AVcc => w.refs().avcc(),
+                    ReferenceVoltage::Internal => w.refs().internal(),
+                });
+                                                
+            }
+
+            fn disable(&self) {
+                self.peripheral.adcsra.reset();
+            }
+
+            fn set_reading(&mut self, channel: $ID, state: bool) {
+                match channel {
+                    $( $ChannelIDPat => self.$pxi = state, )+
+                    _ => unreachable!(),
+                }
+            }
+
+            fn is_reading(&self, channel: $ID) -> bool {
+                match channel {
+                    $( $ChannelIDPat => self.$pxi, )+
+                    _ => unreachable!(),
+                }
+            }
+
+            pub fn release(self) -> $ADC {
+                self.disable();
+                self.peripheral
+            }
+        }
+
+        impl<WORD, PIN> OneShot<$Adc, WORD, PIN> for $Adc
+        where
+            WORD: From<u16>,
+            PIN: Channel<$Adc, ID=$ID>,
+        {
+            type Error = ();
+
+            fn read(&mut self, _pin: &mut PIN) -> nb::Result<WORD, Self::Error> {
+                let channel = PIN::channel();
+
+                match (self.is_reading(channel), self.peripheral.adcsra.read().adsc().bit_is_set()) {
+                    (true, true) =>  Err(nb::Error::WouldBlock),
+                    (true, false) => {
+                        self.set_reading(channel, false);
+                        Ok(self.peripheral.adc.read().bits().into())
+                    },
+                    (false, _) => {
+                        self.set_reading(channel, true);
+                        self.peripheral.admux.modify(|_, w| w.mux().variant(channel));
+                        self.peripheral.adcsra.modify(|_, w| w.adsc().set_bit());
+                        Err(nb::Error::WouldBlock)
+                    }
+                }
+            }
+        }
+
+        $(
+            impl Channel<$Adc> for $PXi<Analog> {
+                type ID = $ID;
+                fn channel() -> Self::ID {
+                    $ChannelIDExpr
+                }
+            }
+
+            impl<MODE> $PXi<MODE> {
+                    /// Make this pin a analog input and enable the internal pull-up
+                    pub fn into_analog_input(self, adc: &mut $Adc) -> $PXi<Analog> {
+                        adc.peripheral.didr0.modify(|_, w| w.$name().set_bit());
+                        $PXi { _mode: core::marker::PhantomData }
+                    }
```

Comment:
This is what I'd do as well, so :+1:.  Though, looking into the future, ATmega32U4 has 2 _Digital Input Disable Registers_ so if you want, I would suggest making the register name a macro parameter as well.

Reply:
Ok changed that.